src/hibernate-resume/hibernate-resume-generator.c

```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
         return 0;
 }
 
+static int parse_efi_hibernate_location(void) {
+        int r = 0;
+
+#if ENABLE_EFI
+        static const JsonDispatch dispatch_table[] = {
+                { "uuid",         JSON_VARIANT_STRING,   json_dispatch_id128,        offsetof(EFIHibernateLocation, id),            JSON_MANDATORY             },
+                { "offset",       JSON_VARIANT_UNSIGNED, json_dispatch_uint64,       offsetof(EFIHibernateLocation, offset),        JSON_MANDATORY             },
+                { "id",           JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, id),            JSON_PERMISSIVE|JSON_DEBUG },
+                { "imageId",      JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, image_id),      JSON_PERMISSIVE|JSON_DEBUG },
+                { "versionId",    JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, version_id),    JSON_PERMISSIVE|JSON_DEBUG },
+                { "imageVersion", JSON_VARIANT_STRING,   json_dispatch_const_string, offsetof(EFIHibernateLocation, image_version), JSON_PERMISSIVE|JSON_DEBUG },
+                {},
+        };
+
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ char *location_str = NULL, *device = NULL, *id = NULL, *image_id = NULL,
+                       *version_id = NULL, *image_version = NULL;
+        char node[STRLEN("UUID=") + SD_ID128_UUID_STRING_MAX];
+        EFIHibernateLocation location = {};
+        int log_level, log_level_ignore;
+
+        log_level = arg_resume_device ? LOG_DEBUG : LOG_ERR;
+        log_level_ignore = arg_resume_device ? LOG_DEBUG : LOG_WARNING;
+
+        r = efi_get_variable_string(EFI_SYSTEMD_VARIABLE(HibernateLocation), &location_str);
+        if (r == -ENOENT) {
+                log_debug("EFI variable HibernateLocation is not set, skipping.");
+                return 0;
+        }
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to get EFI variable HibernateLocation: %m");
+
+        r = json_parse(location_str, 0, &v, NULL, NULL);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to parse HibernateLocation JSON object: %m");
+
+        r = json_dispatch(v, dispatch_table, NULL, JSON_LOG, &location);
+        if (r < 0)
+                return r;
+
+        xsprintf(node, "UUID=" SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(location.uuid));
+
+        device = fstab_node_to_udev_node(node);
+        if (!device)
+                return log_full_errno(log_level, SYNTHETIC_ERRNO(ENODEV), "Failed to get resume device through HibernateLocation.");
```

Comment:
just do:

```c
if (asprintf(&device, "/dev/disk/by-uuid/" SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(location.uuid)) < 0)
```

instead of first formatting this as fstab UUID= expressoin, and then converting it to /dev/disk/by-uuid/ syntax